File: SegueingInfo/SegueingInfo.m
//
//  SegueingInfo.m
//  SegueingInfo
//
//  Created by Ian on 8/17/14.
//  Copyright (c) 2014 Adorkable. All rights reserved.
//

#import "SegueingInfo.h"

#if USE_SWIZZLING
#import <RSSwizzle/RSSwizzle.h>
#endif

#if TARGET_OS_IPHONE
@implementation UIViewController (SegueingInfo)
#else
@implementation NSViewController (SegueingInfo)
#endif

/**
 *  If swizzling is enabled set up swizzling
 *
 */
#if USE_SWIZZLING
+ (void)load
{
    static const void *key = &key;
    
    SEL selector = @selector(prepareForSegue:sender:);
    
#if TARGET_OS_IPHONE
    RSSwizzleInstanceMethod(self,
                            selector,
                            RSSWReturnType(void),
                            RSSWArguments(UIStoryboardSegue *segue, id sender),
                            RSSWReplacement(
    {
        RSSWCallOriginal(segue, sender);
        
        [UIViewController prepareDestinationViewControllerForSegue:segue withInfo:sender];

    }), RSSwizzleModeOncePerClass, key);
#else
    RSSwizzleInstanceMethod(self,
                            selector,
                            RSSWReturnType(void),
                            RSSWArguments(NSStoryboardSegue *segue, id sender),
                            RSSWReplacement(
    {
        RSSWCallOriginal(segue, sender);
        
        [NSViewController prepareDestinationViewControllerForSegue:segue withInfo:sender];
        
    }), RSSwizzleModeOncePerClass, key);
#endif
}
#endif

/**
 *  Prepare compatible self as destination
 *
 *  @param segue Segue
 *  @param info  Info to pass
 */
- (void)prepareAsDestinationViewControllerForSegue:(StoryboardSegueClass *)segue withInfo:(id)info
{
    if ( [self conformsToProtocol:@protocol(SegueingInfoProtocol) ] )
    {
        id<SegueingInfoProtocol> selfAsProtocol = (id<SegueingInfoProtocol>)self;
        if ( [selfAsProtocol respondsToSelector:@selector(destinationPrepareForSegue:info:) ] )
        {
            [selfAsProtocol destinationPrepareForSegue:segue info:info];
        }
    }
}

/**
 *  Prepare compatible segue destination
 *
 *  @param segue Segue
 *  @param info  Info to pass
 */
+ (void)prepareDestinationViewControllerForSegue:(StoryboardSegueClass *)segue withInfo:(id)info
{
    id destination = nil;
#if TARGET_OS_IPHONE
    destination = segue.destinationViewController;
#else
    destination = segue.destinationController;
#endif
    
    if ( [destination respondsToSelector:@selector(prepareAsDestinationViewControllerForSegue:withInfo:) ] )
    {
        [destination prepareAsDestinationViewControllerForSegue:segue withInfo:info];
    }
}

@end

#if !USE_SWIZZLING

@implementation SegueingInfoViewController

/**
 *  Inhereted forwarding to SegueingInfo handlers for subclasses of SegueingInfoViewController
 *
 *  @param segue  Segue
 *  @param sender Sender
 */
- (void)prepareForSegue:(StoryboardSegueClass *)segue sender:(id)sender
{
    [super prepareForSegue:segue sender:sender];
    
    [ViewControllerClass prepareDestinationViewControllerForSegue:segue withInfo:sender];
}

@end

#endif

#if TARGET_OS_IPHONE
@implementation UINavigationController (SegueingInfo)

/**
 *  Pop View Controller
 *
 *  @param animated Animated or not
 *  @param info     Info to pass
 *
 */
- (UIViewController *)popViewControllerAnimated:(BOOL)animated info:(id)info
{
    if (self.viewControllers.count > 1 && [self.viewControllers[self.viewControllers.count - 2] isKindOfClass:[UIViewController class] ] )
    {
        UIViewController *topViewController = (UIViewController *)self.viewControllers[self.viewControllers.count - 2];
        [topViewController prepareAsDestinationViewControllerForSegue:nil withInfo:info];
    }
    return [self popViewControllerAnimated:animated];
}

/**
 *  Pop to specified View Controller
 *
 *  @param viewController View Controller to pop to
 *  @param animated       Animated or not
 *  @param info           Info to pass
 *
 */
- (NSArray *)popToViewController:(UIViewController *)viewController animated:(BOOL)animated info:(id)info
{
    [viewController prepareAsDestinationViewControllerForSegue:nil withInfo:info];
    return [self popToViewController:viewController animated:animated];
}

/**
 *  Pop to Root View Controller
 *
 *  @param animated Animated or not
 *  @param info     Info to pass
 *
 */
- (NSArray *)popToRootViewControllerAnimated:(BOOL)animated info:(id)info
{
    if (self.viewControllers.count > 0 && [self.viewControllers.firstObject isKindOfClass:[UIViewController class] ] )
    {
        UIViewController *rootViewController = (UIViewController *)self.viewControllers.firstObject;
        [rootViewController prepareAsDestinationViewControllerForSegue:nil withInfo:info];
    }
    return [self popToRootViewControllerAnimated:animated];
}

@end
#endif

File: Sources/SegueingInfo.swift
//
//  SegueingInfo.swift
//  SegueingInfo
//
//  Created by Ian Grossberg on 11/30/15.
//  Copyright © 2015 Adorkable. All rights reserved.
//

#if os(OSX)
    import Cocoa
#elseif os(iOS)
    import UIKit
#endif

#if os(OSX)
    public typealias SegueingInfoViewControllerClass = NSViewController
    public typealias SegueingInfoStoryboardSegueClass = NSStoryboardSegue
#elseif os(iOS)
    public typealias SegueingInfoViewControllerClass = UIViewController
    public typealias SegueingInfoStoryboardSegueClass = UIStoryboardSegue
#endif

public typealias SegueingInfoInfoClass = AnyObject

public extension SegueingInfoViewControllerClass {

    /**
    *  Static Utility selector that handles passing information to a waiting and able destination View Controller. Usually called from prepareForSegue:sender:
    *
    *  @param segue The segue being performed
    *  @param withInfo  The information being passed to the destination of the segue
    */
    public class func prepareDestinationViewControllerForSegue(_ segue : SegueingInfoStoryboardSegueClass, withInfo info : SegueingInfoInfoClass) {
        
        #if os(OSX)
            segue.destinationController.treatAsDestinationViewController(segue, withInfo: info)
        #elseif os(iOS)
            segue.destination.treatAsDestinationViewController(segue, withInfo: info)
        #endif
    }
    
    /**
     *  Utility selector that handles passing information to a waiting and able destination View Controller. Usually called from prepareForSegue:sender:
     *
     *  @param segue The segue being performed
     *  @param withInfo  The information being passed to the destination of the segue
     */
    public func prepareDestinationViewControllerForSegue(_ segue : SegueingInfoStoryboardSegueClass, withInfo info : SegueingInfoInfoClass) {
        type(of: self).prepareDestinationViewControllerForSegue(segue, withInfo: info)
    }
}

public extension SegueingInfoViewControllerClass {
    /**
     *  Prepare compatible self as destination
     *
     *  @param segue Segue
     *  @param withInfo  Info to pass
     */
    internal func treatAsDestinationViewController(_ segue : SegueingInfoStoryboardSegueClass?, withInfo info : SegueingInfoInfoClass) {

        // Note: We aren't extending UINavigationController or UITabBarController, this way functionality is more explicitly controlled by clients of the library
        // By default we'll seek through a NavigationControllerClass to it's root or UITabBarController to it's first VC
        // But if the client explicity conforms to SegueingInfoDestination he/she can decide what do to and potentially implement more complicated behavior
        if let destination = self as? SegueingInfoDestination {
            destination.destinationPrepareForSegue(segue, withInfo: info)
        } else {
            #if os(iOS)
                if let navigationController = self as? UINavigationController,
                    let firstViewController = navigationController.viewControllers.first {
                        firstViewController.treatAsDestinationViewController(segue, withInfo: info)
                } else
                {
                    if let tabBarController = self as? UITabBarController {
                        if let selectedViewController = tabBarController.selectedViewController {
                            
                            selectedViewController.treatAsDestinationViewController(segue, withInfo: info)
                            
                        } else if let viewControllers = tabBarController.viewControllers {
                            if viewControllers.count > 0 {
                                viewControllers[0].treatAsDestinationViewController(segue, withInfo: info)
                            }
                        }
                    }
                }
            #endif
        }
    }
}

/**
 *  Protocol for receiving passed information
 */
public protocol SegueingInfoDestination : NSObjectProtocol {
    
    /**
     *  Destination View Controller that receives information passed from source View Controller
     *
     *  @param segue The segue being performed
     *  @param withInfo  The information being passed from the source of the segue
     */
    func destinationPrepareForSegue(_ segue : SegueingInfoStoryboardSegueClass?, withInfo info : SegueingInfoInfoClass)
}

/// Backwards compatability with < 2.0.0 versions of Framework
public typealias SegueingInfoProtocol = SegueingInfoDestination

open class SegueingInfoViewController : SegueingInfoViewControllerClass {
    
    override open func prepare(for segue: SegueingInfoStoryboardSegueClass, sender: Any?) {
        super.prepare(for: segue, sender: sender)
        
        if let info = sender {
            self.prepareDestinationViewControllerForSegue(segue, withInfo: info as SegueingInfoInfoClass)
        }
    }
}

